File: chapter2/chapter2.cpp
#include <iostream>
#include <math.h>
#include <vector>
#include <array>
#include <algorithm>
#include <cstdlib>
#include <stdio.h> 
#include <stdlib.h>
#include <tuple>
#include <exception>
#include <memory>
#include <chrono>
#include <mutex>
#include <future>
#include <iomanip>
#include <string.h>
#include "chapter2.h"


bool leap_year(unsigned year)
{ 
  if (year % 400 == 0)
    return true;
  if (year % 100 == 0)
    return false;
  if (year % 4 == 0)
    return true;
  return false;
}
	
	
namespace chapter2
{
	
unsigned sigma(unsigned num)
{
  unsigned sum = num;
  for ( ; num-- ; ) {
    sum += num;
  }
  return sum;
}
	
unsigned factorial(unsigned num) 
{
  unsigned result = 1; 
  for (; num > 0 ; num--) {
    result *= num;
  }
  return result; 
}
	
void Star_Art::drawLeftStars(unsigned count, const char & c)
{
  for (int i = 0; (i < max_len); i++) {
    if (count-- > 0) {
      printf("%c", c);
    }
  }
}

void Star_Art::drawRightStars(unsigned count, const char &c)
{
  for (int i = max_len; i > 0; i--) {
    if (i > count)
      printf(" ");
    else
      printf("%c", c);
    }
}

void Star_Art:: drawCenteredStars(unsigned count, const char &c)
{
  std::vector<char> data;
  int l = std::floor(max_len / 2);
  int r = std::floor(max_len / 2) + 1;
  data.resize(max_len, '\0');

  while ( count-- > 0) {
    if (l-- >= 0)
      data[l] = c;
    if (r++ >= 75)
      data[r] = c;
  }

  for (auto x : data) {
    std::cout << x;
  }
}

void Star_Art::drawLeftChars(unsigned count, const char& c) 
{
  drawLeftStars(count, c);
}
void Star_Art::drawRightChars(unsigned count, const char& c) 
{
  drawRightStars(count, c);
}
void Star_Art::drawCenteredChars(unsigned count, const char& c) 
{
  drawCenteredStars(count, c);
}

unsigned threes_and_fives() 
{
  unsigned sum = 0;
  for (int i = 100; i <= 4000000; i++) {
    if ((i % 3 == 0) ^ (i % 5 == 0)) {
      sum += i;
    }
  }
  return sum;
}

unsigned better_threes_and_fives(unsigned start, unsigned end)
{
  unsigned sum = 0;
  for (int i = start; i <= end; i++) {
    if ((i % 3 == 0) ^ (i % 5 == 0)) {
      sum += i;
    }
  }
  return sum;
}

void generate_coin_chnage(unsigned cents)
{
  unsigned quarters = 0, dimes = 0, nickels = 0, pennies = 0;
  while (cents > 0) {
    if (cents >= 25) {
      cents -= 25;
      quarters++;
    } else if (cents >= 10) {
      cents -= 10;
      dimes++;
    } else if (cents >= 5) {
      cents -= 5;
      nickels++;
    } else if (cents >= 1) {
      cents -= 1;
      pennies++;
    }
  }

  printf("quarters-[%d]  dimes-[%d]  nickels-[%d]  pennies-[%d]",
  quarters,
  dimes,
  nickels,
  pennies);
}

double messy_math_mashup(double num) {
  double sum = 0;
  for (int count = 0; count <= num; count++) {
    if (count % 3 == 0)
      continue;
    else if (count % 7 == 0)
      sum += static_cast<double>(2*count);
    else if (count == num / 3)
      return -1;
    else
      sum += count;
  }
  return sum;
}

void twelve_bar_blues() {
  for (int i = 12; i--; ) {
    printf("[%d] chick bool chick\n", i);
    std::this_thread::sleep_for(std::chrono::seconds(1));
  }
}

unsigned fibonacci(unsigned index) 
{
  unsigned reg[2] = { 0, 1 };
  unsigned tmp;

  for (int i = 0; i <= index; i++) {
    if (i > 1) {
      tmp = reg[1];
      reg[1] = reg[1] + reg[0];
      reg[0] = tmp;
    }
  }
  return reg[+(index > 0)];
}

unsigned sumToOne(unsigned num)
{
  unsigned sum = 0;
  while (num > 0) {
    sum += num % 10;
    num /= 10;
  }

  if (sum >= 10) {
    return sumToOne(sum);
  }
}

void clock_handle_angles(unsigned seconds)
{
/*
  60 sec = 1min = 6 degs
  1 hour = 30 degs
*/

  double sec = seconds % 60;
  double min = (seconds / 60) % 60;
  double hours = (seconds/ 3600) % 12 ;
  double extra_sec;
  double extra_min;

  double angle_sec, angle_hour, angle_min;

  extra_sec = 6.0 * (sec / 60.0); /*  calc extra seconds in current minute */
  extra_min = 30.0 * (min / 60.0); // /* calc extra minutes in current hour */

/* conversion from time to angle */
  angle_sec = (sec / 60) * 360.0; 
  angle_min = ((min / 60) * 360) + extra_sec;
  angle_hour = ((hours / 12) * 360) + extra_min;;

  printf(" Hour_Hand %f degs  Min_Hand %f degs  Second_Hand %f degs \n", angle_hour, angle_min, angle_sec);
}

bool isPrime(int number) {
  for (int i = 2; i <= std::floor(std::sqrt(number)); i++) {
    if (number % i == 0) {
      return false; 
    }
  }
  return true;
}

double rockin_the_dojo_sweatshirt(unsigned number)
{
  double cost = 20;

  if (number < 2)
    return cost;

  switch (number) {
    case 2:
      return 0.09 * 20;
    case 3:
      return 0.19 * 20;
    default:
      return 0.35 * 20;
  }
}

unsigned extract_o_matic(double num, signed digitNum)
{
  unsigned digit = 0;

  if (digitNum < 0) {
    for (int i = digitNum; i++; ) {
      num *= 10.0;
      printf("num %f \n", num);
    }
  }

  if (digitNum >= 0) {
    for (int i = digitNum; i--; ) {
      num /= 10.0;
      printf("num %f \n", num);
    }
  }

  digit = static_cast<int>(std::abs(num)) % 10;
  return digit;
}

unsigned most_significant_digit(double number)
{
  unsigned digit = 0;
  while (std::abs(number) > 10) {
    number = number / std::pow(10, 1);
    digit = static_cast<int>(std::abs(number)) % 10;
  }
  return digit;
}

unsigned GamingFundamentals::rollOne()
{
  return std::rand() % 6 + 1;
}

void GamingFundamentals::playFives(unsigned num)
{
  unsigned data; 

for (; num--;) {
  data = rollOne();
  printf("rolled value %d \n", data);
  if (data == 5)
    printf("\t That's good luck\n");
  }
}

void GamingFundamentals::playStatistics()
{
  unsigned min = 7, max = 0, data = 0;
  for (int i = 8; i--; ) {
    data = rollOne();
  if (data > max)
    max = data; 
  if (data < min)
    min = data;
  }
	
  printf("min {%d}  max {%d}", min, max);
}

void GamingFundamentals::playStatistics2()
{
  unsigned min = 7, max = 0, data = 0, sum = 0;
  for (int i = 8; i--; ) {
    data = rollOne();
    if (data > max)
      max = data;
    if (data < min)
      min = data;
    sum += data;
  }
  printf("min {%d}  max {%d}  total-sum {%d}", min, max, sum);
}

void GamingFundamentals::playStatistics4(unsigned num)
{
  unsigned min = 7, max = 0, data = 0, sum = 0, average = 0;
	
  for (int i = num; i--; ) {
    data = rollOne();
    if (data > max)
      max = data;
    if (data < min)
      min = data;
    sum += data;
  }
	
  if (num) {
    average = sum / num;
  }
	
  printf("min {%d}  max {%d}  total-sum {%d}  avg {%d}", min, max, sum, average);
}

void GamingFundamentals::statistics_until_doubles()
{
  unsigned min = 7, max = 0, data = 0, sum = 0, average = 0;
  std::map<unsigned, unsigned> stat;
  int i = 1;

  for (;;i++) {
    data = rollOne();
	  
    if (stat.find(data) == stat.end()  ) {
      stat.insert({data, 1});
    } else {
      break;
    }
	  
    if (data > max)
      max = data;
    
    if (data < min)
      min = data;

    sum += data;
  }
  average = sum / i;
  printf("min {%d}  max {%d}  total-sum {%d}  avg {%d} tick {%d}", min, max, sum, average, i);
}

WhereIsClaire WhereIsClaire::reset()
{
  pos[0] = 1;
  pos[1] = 1;
  return *this;
}

WhereIsClaire WhereIsClaire::moveBy(signed x, signed y)
{
  pos[0] += x;
  pos[1] += y;
  return *this;
}

signed WhereIsClaire::xLocation()
{
  return pos[0];
}

signed WhereIsClaire::yLocation()
{
  return pos[1];
}


std::string  Date_on_Deserted_Island::weekdayName(unsigned weekdayNum)
{
  std::string day{};

  if ((weekdayNum < 1) || (weekdayNum > 365))
    return day; 

  switch (weekdayNum % 7) {
  case 1:
    day = "Sunday";
    break;
  case 2:
    day = "Monday";
    break;
  case 3:
    day =  "Tuesday";
    break;
  case 4:
    day = "Wednesday";
    break;
  case 5:
    day =  "Thurday";
    break;
  case 6:
    day = "Friday";
    break;
  case 0:
    day = "Saturday";
    break;
  }
	
  return day; 
}

void Date_on_Deserted_Island::someDays ()
{
  std::string day;

  for (int i = 17; i--; ) {
    day = weekdayName( (rand() % 7) + 1);
    printf("%s\n", day.c_str());
    if (day != "Sunday" && day != "Saturday") {
      printf("\t WORK HARD\n");
    }
  }
}

const char* Date_on_Deserted_Island::monthName(unsigned monthNum)
{
  try {
    if (monthNum > 12 || monthNum < 1)
      throw "invalid month";
    return std::get<1> (days_in_month_map[monthNum - 1]);
  } catch (...) {
    return "";
  }
}

unsigned Date_on_Deserted_Island::month_to_day_in_2017(unsigned monthNum)
{
try {
   if ((monthNum < 1 || monthNum > 12))
    throw "invalid month";
  return std::get<0>(days_in_month_map[monthNum - 1]);
  } catch (...) {
    return 0;
  }
}

const char * Date_on_Deserted_Island::day_to_month(unsigned dayNum)
{
  unsigned start = 0, end = 0;
  
  if ((dayNum < 0 || dayNum> 365))
    return "";
  
  for (int i = 0; i < days_in_month_map.size(); i++) {
    end += std::get<0> (days_in_month_map[i]);
    if (dayNum > start && dayNum <= end) {
      return std::get<1>(days_in_month_map[i]);;
    }
    start = end;
  }
  return "";
}

void Date_on_Deserted_Island::reset_map()
{
  const std::array<std::tuple<int, int>, 12> data = {
    std::make_pair(31, 0),
    std::make_pair(28, 3),
    std::make_pair(31, 3),
    std::make_pair(30, 6),
    std::make_pair(31, 1),
    std::make_pair(30, 4),
    std::make_pair(31, 6),
    std::make_pair(31, 2),
    std::make_pair(30, 5),
    std::make_pair(31, 0),
    std::make_pair(30, 3),
    std::make_pair(31, 5)
  };

  for (int i = 0; i < 12; i++) {
    std::get<0>(days_in_month_map[i]) = std::get<0>(data[i]);
    std::get<2>(days_in_month_map[i]) = std::get<1>(data[i]);
  }
  year = 2017;
}

std::string Date_on_Deserted_Island::full_data(unsigned dayNum)
{
  const char *month = day_to_month(dayNum);
  unsigned offset;
  int i;

  for (i = 0; i < days_in_month_map.size() && dayNum > std::get<0>(days_in_month_map[i]) ; i++) {
    dayNum -= std::get<0>(days_in_month_map[i]);
  }
  offset = std::get<2>(days_in_month_map[i]);

  return weekdayName( offset + dayNum) \
  + " " + month + \
  + " " + std::to_string(dayNum) + \
  + " " + std::to_string(year); 
}

void Date_on_Deserted_Island::offset_calendar(unsigned val, std::tuple<unsigned, const char*, unsigned> &tuple)
{
  std::get<2>(tuple) = (std::get<2>(tuple) + val) % 7;
}

std::string Date_on_Deserted_Island::full_data2(unsigned dayNum)
{
  bool leaped = 0;
  std::string result{};

  while (dayNum > 365) {
    for (int i = 0; i < days_in_month_map.size(); i++) {
      if (leaped) {
        if (std::get<0>(days_in_month_map[i]) == 29)  /* remove shift 29th day from feburary */
          std::get<0>(days_in_month_map[i]) = 28;
        if (i <= 1)   /* shift jan-feb due to previous leap */
          offset_calendar(1, days_in_month_map[i]);
      }
      offset_calendar(1, days_in_month_map[i]);
    }

    dayNum -= leaped ? 366 : 365;
    if (leaped) {
      leaped = false;
    }

    if (leap_year(++year)) { /* leaped year */
      leaped = true;
      for (int i = 0; i < days_in_month_map.size(); i++) {
        if (i == 1) 
          std::get<0>(days_in_month_map[i]) = 29;
        if (i >= 2)   /* shift march-dec by one due to add leap year */
          offset_calendar(1, days_in_month_map[i]);
      }
    }
  }

  for (int i = 1; i <= 12; i++) {
    view_calendar_month(i);  /* print calendar year */
  }

  result = full_data(dayNum);
  reset_map();
  return result;
}

void Date_on_Deserted_Island::view_calendar_month(unsigned month)
{
  std::tuple<unsigned, const char*, unsigned> tuple;
  unsigned index = 0, count = 1, start_day = 0, end_day = 0;

  if (month < 1 || month > 12) {
    return;
  }

  fill_calendar(month);  // update matrix 
  tuple = days_in_month_map[month - 1]; // readin data for mapping calendar
  printf("\t%s\n", std::get<1>(tuple)); // print calendar month to view 

  for (int k = 1; k <= 7; k++) {
    printf("[%s]", days_of_week[k%7]);
  if (k == 7) 
    printf("\n");
  }

  start_day = std::get<2>(tuple);
  end_day = (std::get<0>(tuple) + std::get<2>(tuple));

  for (auto r = 0; r < 6; r++) {
    for (auto c = 0; c < 7; c++) {
      index = (r * 7) + c;
      if (index >= start_day && index < end_day) {
        (count < 10) ? printf("[**%d]", count) : printf("[*%d]", count);
        count++;
      } else {
        printf("[***]");
      }
    }
    printf("\n");
  }
}

void Date_on_Deserted_Island::fill_calendar(unsigned month)
{
  unsigned row = 0, col = 0;
  std::tuple<unsigned, const char*, unsigned> tuple = days_in_month_map[month - 1];

  clear_calendar();
  for (int i = std::get<2>(tuple); i < std::get<0>(tuple); i++) {
    row = std::floor(i / 7);
    col = (i % 7);
    curr_calendar[row][col] = days_of_week[col];
  }
}

void Date_on_Deserted_Island::clear_calendar()
{
  for (auto r = 0; r < 6; r++) {
    for (auto c = 0; c < 7; c++) {
      curr_calendar[r][c] = "*";
    }
  }
}

	
	
};
	
	
	

File: chapter2/chapter2.py
import math
import random

def sigma (num):
  sum = 0
  for n in range(1, num + 1):
    sum += n
  return sum

def factorial (num):
  sum = 1
  for n in range(1, num + 1):
    sum *= n
  return sum

class StarArt():
  def __init__(self):
    self.length = 75
  def draw_left_stars(self, num):
    result = ''
    while self.length is not  0 and num > 0:
      num -= 1
      self.length -= 1;
      result += '*'
    self.length = 75
    return result
  def draw_right_stars(self, num):
    start_point = self.length - num
    counter = 0
    result = ''

    while start_point > 0  and  self.length is not 0:
      if counter >= start_point  :
        result += '*'
      else:
        result += ' '
      self.length -= 1
      counter += 1
    self.length = 75
    return result

  def draw_center_stars(self, num, character = '*'):
    result = ''
    counter = 1
    mid = 37
    length_from_mid =  int(math.floor(num * 0.5)) + ( 0 if num % 2 == 1 else - 1 )
    start = mid - length_from_mid
    end = start + num - 1

    if num <= 0:
      return result


    while self.length :
      if counter >= start and counter <= end:
        result += character
      else:
        result += ' '
      counter += 1
      self.length -= 1

    return result

def threes_and_fives(start = 100, end = 4000000):
  sum = 0
  for i in range(start, end + 1):
    if (not i % 3) ^ (not i % 5):
      sum += i
  return sum

def generate_coin_change (number):
  coins = []

  while number :

    if number >= 100:
      number -= 100
      coins.append('dollar')
    elif number >=50:
      number -=50
      coins.append('half-dollar')
    elif number >= 25:
      number -= 25
      coins.append('quarter')
    elif number >= 10:
      number -= 10
      coins.append('dime')
    elif number >= 5:
      number -= 5
      coins.append('nickel')
    elif number >=1:
      number -= 1
      coins.append('penny')
  return coins

def messyMath(num):
  sum = 0
  for i in range (0, num +1 ):
    if  not i % 3:
      pass
    elif not i % 7:
      sum += 2 * i
    elif num/3 is num:
      return -1
    else:
      sum += i
  return sum

def twelve_bar_blues():
  for i in range(0, 12):
    print(str(i) + 'chick' + 'boom' + 'chick')

def fibonacci(number):
  reg = [0, 1]
  result = 0

  if number < 0:
    return None

  if number < 2:
    return number

  for i in range(1, number + 1 ):
    result = reg[1]
    reg[0], reg[1] = reg[1], reg[0]
    reg [1] = reg[0] + reg[1]

  return result

def sum_to_one_digit(number):
  while number > 9:
    tmp = number
    number = 0
    while tmp >= 1:
      number += tmp % 10
      tmp = math.floor(tmp / 10)
  return number

def clock_hand_angles(seconds):
  s = seconds % 60
  m = (seconds / 60) % 60
  h = (seconds / 3600) % 12
  print(s, m, h)
  # 60 second == 6 degrees (i.e. minute hand)
  # 60 min  == 30 degree (i.e.  hour hand)

  extra_m_angle = (s / 60.0) * 6
  extra_h_angle =  (m / 60.0) * 30

  s_hand = ((s / 60.0) * 360.0)
  m_hand = ((m / 60.0) * 360.0) + extra_m_angle
  h_hand = ((h / 12.0) * 360.0) + extra_h_angle

  return {'second_hand': s_hand, 'minute_hand': m_hand, 'hour_hand': h_hand }

def is_prime (number):
  for i in range (2, int(math.sqrt(number + 1)) + 1):
    if not number % i:
      return False
  return True

def rockin_dojo_sweatshirt(num):
  amount = 20
  num  = num if num < 4 else 3

  if num < 0:
    return None

  return {
    0: amount,
    1: amount * 0.9,
    2: amount * 0.19,
    3: amount * 0.35,
  }[num]

def extract_o_matic (num, digitNum):
  for i in range(0, digitNum):
    num  /=  10
  return num % 10

def most_sig_digit(number):
  result = 0
  while number:
    result = number % 10
    number /= 10
  return result

class Gaming_Fundamentals():
  def __init__(self):
    pass
  def rollOne(self):
    return (math.floor(random.random() * 6)) + 1
  def playFives(self, number):
    for i in range(0, number):
      data = int(self.rollOne())
      print(data)
      if data is 5:
        print 'good luck!'
  def playStatistics(self):
    hi = None
    low = None

    for i in range(0, 8):
      data = self.rollOne()
      if hi is None or data > hi:
        hi = data
      if low is None or data < low:
        low = data
    return {'low': low, 'hi': hi}

  def playStatistics2(self):
    hi = None
    low = None
    acc = 0
    for i in range(0, 8):
      data = self.rollOne()
      if hi is None or data > hi:
        hi = data
      if low is None or data < low:
        low = data
      acc += data
    return {'low': low, 'hi': hi, 'acc': acc}

  def playStatistics3(self, number):
    hi = None
    low = None
    acc = 0
    for i in range(0, number):
      data = self.rollOne()
      if hi is None or data > hi:
        hi = data
      if low is None or data < low:
        low = data
      acc += data
    return {'low': low, 'hi': hi, 'acc': acc}

  def playStatistics4(self, number):
    hi = None
    low = None
    acc = 0
    avg = 0

    if number <= 0:
      return None

    for i in range(0, number):
      data = self.rollOne()
      if hi is None or data > hi:
        hi = data
      if low is None or data < low:
        low = data
      acc += data
    return {'low': low, 'hi': hi, 'acc': acc, 'avg': acc / number }

def stat_until_double():
  min = None
  max = None
  prev = 0
  sum = 0
  count = 0
  current_role = 0

  while 1:

    current_role = int ((random.random() * 20) + 1 )

    if current_role is prev:
      break

    prev = current_role

    if min is None or current_role < min:
      min = current_role
    if max is None or current_role > max:
      max =  current_role

  sum += current_role
  count += 1
  return {'max': max, 'min': min, 'avg': sum/ count, 'count': count }

class Claire_Is_Where():
  def __init__(self):
    self.x = 0
    self.y = 0

  @property
  def reset(self):
    self.x = 0
    self.y = 0
    return self

  def moveBy(self, x_offset, y_offset):
    self.x += x_offset
    self.y += y_offset
    return self
  def xLocation(self):
    return self.x

  def yLocation(self):
    return self.y

  def current_location(self):
    return {'x': self.x, 'y': self.y}

  def distFromHome(self):
    x_offset = 0
    y_offset = 0
    x = 0
    y = 0
    steps = 0

    if self.x > 0 and self.y > 0:
      x_offset = 1
      y_offset = 1

    if self.x < 0 and self.y > 0:
      x_offset = -1
      y_offset = 1

    if self.x < 0 and self.y < 0:
      x_offset = -1
      y_pffset = -1

    if self.x > 0 and self.y < 0:
      x_offset = 1
      y_offset = -1

    while x is not self.x or   y is not self.y:
      if x is self.x or y is self.y:
        if x is not self.x:
          steps+= abs(x - self.x)
        if y is not self.y:
          steps+= abs(y- self.y)
        break
      else:
        x += x_offset
        y += y_offset
        steps += 1
    return str(steps) + ' units'

class Date_On_Deserted_Island():

  def __init__(self):
    self.year = 2016
    self.months = { 1: 31, 2: 29, 3: 31, 4: 30, 5: 31, 6: 30, 7: 31,8: 31, 9: 30, 10: 31,  11: 30, 12: 31 }
    self.month_code =  ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'June', 'July', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
    self.days = ['Monday', 'Tuesday', 'Wednesday', 'Thurday', 'Friday', 'Saturday', 'Sunday']
    self.day_index = 5
    self.month_index = 12
    self.day = 31

  def weekday_name(self, number):
    if number < 0 or number > 7:
      return None
    index = (self.day_index + 1) % 7
    return self.days[index]

  def weekday_name2(self, number):

    day = self.day
    day_sel = self.day_index
    month = self.month_index
    year = self.year

    if number < 0 or number > 365:
      return None

    while number :

      # ending leap year || end of year

      if month is 12 and day is 31:
        if year == 2016:
          self.months[2] = 28
        year += 1
        month = 1
        day = 0

      # last day of month

      elif day is self.months[month]:
        month += 1
        day = 0

      day += 1
      day_sel = (day_sel + 1) % 7
      number -= 1

    return  {'month': month, 'days_name': self.days[day_sel], 'day': day, 'year':  year}

  def some_days(self):
    for i in range(0, 17):
      data = self.weekday_name2(int(random.random()*365 + 1 ))
      print(data)
      if data['days_name'] in ['Saturday', 'Sunday']:
        print('Enjoy your days off')

  def month_name(self, num):
    return self.month_code[num - 1]

  def month_to_days(self, num):
    return self.months[num - 1]

  def full_data(self, num):
    data = self.weekday_name2(num);
    if data:
      i = data['month']
      print( data['days_name'] + ' ' +  self.month_name(i)  + ' ' + str(data['day']) + ' ' + str (data['year']) )
    return None

  def full_data_2 (self, number, cb = None ):

    day = self.day
    day_sel = self.day_index
    month = self.month_index
    year = self.year

    if number < 0 :
      return None

    while number :
      if month is 12 and day is 31:
        if cb((year + 1)) if cb else (year + 1) % 4 == 0:
          self.months[2] = 29
        if cb(year) if cb else not year % 4 :
          self.months[2] = 28
        year += 1
        month = 1
        day = 0
      elif day is self.months[month]:
        month += 1
        day = 0
      day += 1
      day_sel = (day_sel + 1) % 7
      number -= 1

    return  {'month': month, 'days_name': self.days[day_sel], 'day': day, 'year':  year}


  def full_data_3 (self, number):
    data =  self.full_data_2(number, is_leap_year)
    if data:
      i = data['month']
      return   '' +  data['days_name'] + ' ' +  self.month_name(i)  + ' ' + str(data['day']) + ' ' + str (data['year'])
    return None

def is_leap_year (year):
  if not year % 400:
    return True
  if not year % 100:
    return False
  if not year % 4:
    return True
  return False

#Test function/method
island = Date_On_Deserted_Island()
print(island.full_data_3( 139947))
